Extension/src/LanguageServer/settings.ts

```diff
@@ -452,6 +461,224 @@ export class CppSettings extends Settings {
+        }
+        vscode.workspace.applyEdit(edits).then(() => vscode.window.showTextDocument(document));
+    }
+
+    public async generateEditorConfig(): Promise<void> {
+        let document: vscode.TextDocument;
+        if (this.resource) {
+            // If a folder is open and '.editorconfig' exists at the root, use that.
+            const uri: vscode.Uri = vscode.Uri.joinPath(this.resource, ".editorconfig");
+            const edits: vscode.WorkspaceEdit = new vscode.WorkspaceEdit();
+            edits.createFile(uri, { ignoreIfExists: true, overwrite: false });
+            try {
+                await vscode.workspace.applyEdit(edits);
+                document = await vscode.workspace.openTextDocument(uri);
+            } catch (e) {
+                document = await vscode.workspace.openTextDocument();
+            }
+        } else {
+            document = await vscode.workspace.openTextDocument();
+        }
+        this.populateEditorConfig(document);
+    }
+
+    // If formattingEngine is set to "Default", searches for .editorconfig with vcFormat
+    // entries, or a .clang-format file, to determine which settings to use.
+    // This is intentionally not async to avoid races due to multiple entrancy.
+    public useVcFormat(document: vscode.TextDocument): boolean {
+        if (this.formattingEngine !== "Default") {
+            return this.formattingEngine === "vcFormat";
+        }
+        if (this.clangFormatStyle !== "file") {
+            // If a clang-format style other than file is specified, don't try to switch to vcFormat.
+            return false;
+        }
+        const cachedValue: boolean | undefined = cachedEditorConfigLookups.get(document.uri.fsPath);
+        if (cachedValue !== undefined) {
+            return cachedValue;
+        }
+        let foundEditorConfigWithVcFormatSettings: boolean = false;
+        const findConfigFile: (parentPath: string) => boolean = (parentPath: string) => {
+            const editorConfigPath: string = path.join(parentPath, ".editorconfig");
+            if (fs.existsSync(editorConfigPath)) {
+                const editorConfigSettings: any = getEditorConfigSettings(document.uri.fsPath);
+                const keys: string[] = Object.keys(editorConfigSettings);
+                for (let i: number = 0; i < keys.length; ++i) {
+                    if (keys[i].startsWith("cpp_")) {
+                        foundEditorConfigWithVcFormatSettings = true;
+                        const didEditorConfigNotice: PersistentState<boolean> = new PersistentState<boolean>("Cpp.didEditorConfigNotice", false);
+                        if (!didEditorConfigNotice.Value) {
+                            vscode.window.showInformationMessage(localize("editorconfig.default.behavior",
+                                "Code formatting is using settings from .editorconfig instead of .clang-format. For more information, see the documentation for the 'C_Cpp.formatting.Default' setting."));
```

Comment:
The "Default" is technically a value and not a setting...not sure is worth clarifying or not.